### ISSUE TITLE
Bug in _collectWarnings in trial

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4748]
|[<img alt="fijal's avatar" src="https://avatars.githubusercontent.com/u/85944?s=50" width="50" height="50">](https://github.com/fijal)| @fijal reported|
|-|-|
|Trac ID|trac#4748|
|Type|defect|
|Created|2010-11-24 12:00:06Z|
|Branch|https://github.com/twisted/twisted/tree/collectwarnings-dict-iteration-4748|

`_collectWarnings` iterates over sys.modules and does a `__setattr__` on them. Some modules in sys.modules might not be actual objects, but arbitrary objects (like `_ModuleProxy` from twisted) that might do stuff on `__setattr__` importing more modules and ending up with a confusing "Dictionary size changed during iteration" message. Can we use .values() instead of .itervalues()?

Attachments:

* [4748.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/482/482b4063693b700098357647f928b1baaa20c430/f4fa805cf6cbbe6ff8a1aa7ffea38d2cfd7b9603.diff) (2913 bytes) - added by fijal on 2011-04-19 12:14:34Z - patch
* [4748.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/482/482b4063693b700098357647f928b1baaa20c430/ae885403d369da5f5ab6dc5d37a8bce2254b8108.patch) (3434 bytes) - added by fijal on 2011-06-10 19:36:29Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4748 4748
type__defect defect
reporter__fijal fijal
priority__normal normal
milestone__ 
branch__branches_collectwarnings_dict_iteration_4748 branches/collectwarnings-dict-iteration-4748
branch_author__fijal fijal
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1290600006000000 1290600006000000
changetime__1307888909000000 1307888909000000
version__None None
owner__exarkun exarkun
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="fijal's avatar" src="https://avatars.githubusercontent.com/u/85944?s=50" width="50" height="50">](https://github.com/fijal)<a name="note_1"></a>|@fijal commented|
|-|-|

oh and an example traceback: http://paste.pocoo.org/show/295415/

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

Example traceback from pastebin link above:

```

(kat)fijal@helmut:~/src/kat-trunk/katstore/katstore$ trial katstore.test.test_server
katstore.test.test_server
  TestStoreServer
    test_historical_data ... Traceback (most recent call last):
  File "/home/fijal/bin/trial", line 22, in <module>
    run()
  File "/home/fijal/src/twisted/trunk/twisted/scripts/trial.py", line 346, in run
    test_result = trialRunner.run(suite)
  File "/home/fijal/src/twisted/trunk/twisted/trial/runner.py", line 796, in run
    return self._runWithoutDecoration(test)
  File "/home/fijal/src/twisted/trunk/twisted/trial/runner.py", line 825, in _runWithoutDecoration
    run()
  File "/home/fijal/src/twisted/trunk/twisted/trial/runner.py", line 820, in <lambda>
    run = lambda: suite.run(result)
  File "/home/fijal/src/twisted/trunk/twisted/trial/runner.py", line 289, in run
    TestSuite.run(self, result)
  File "/home/fijal/src/twisted/trunk/twisted/trial/unittest.py", line 1390, in run
    test(result)
  File "/home/fijal/src/twisted/trunk/twisted/trial/unittest.py", line 1378, in __call__
    return self.run(result)
  File "/home/fijal/src/twisted/trunk/twisted/trial/runner.py", line 165, in run
    super(LoggedSuite, self).run(result)
  File "/home/fijal/src/twisted/trunk/twisted/trial/unittest.py", line 1390, in run
    test(result)
  File "/home/fijal/src/twisted/trunk/twisted/trial/unittest.py", line 1378, in __call__
    return self.run(result)
  File "/home/fijal/src/twisted/trunk/twisted/trial/runner.py", line 138, in run
    test(result)
  File "/home/fijal/src/twisted/trunk/twisted/trial/unittest.py", line 1378, in __call__
    return self.run(result)
  File "/home/fijal/src/twisted/trunk/twisted/trial/runner.py", line 138, in run
    test(result)
  File "/home/fijal/src/twisted/trunk/twisted/trial/unittest.py", line 1378, in __call__
    return self.run(result)
  File "/home/fijal/src/twisted/trunk/twisted/trial/runner.py", line 138, in run
    test(result)
  File "/home/fijal/src/twisted/trunk/twisted/trial/unittest.py", line 745, in __call__
    return self.run(*args, **kwargs)
  File "/home/fijal/src/twisted/trunk/twisted/trial/unittest.py", line 1103, in run
    _collectWarnings(self._warnings.append, runThunk)
  File "/home/fijal/src/twisted/trunk/twisted/trial/unittest.py", line 181, in _collectWarnings
    for v in sys.modules.itervalues():
RuntimeError: dictionary changed size during iteration
```

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_4"></a>|@thijstriemstra commented|
|-|-|

A [similar issue](http://dev.rtmpy.org/ticket/51) existed for RTMPy and Twisted 8.2.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="fijal's avatar" src="https://avatars.githubusercontent.com/u/85944?s=50" width="50" height="50">](https://github.com/fijal)<a name="note_6"></a>|@fijal commented|
|-|-|

Ready for review. According to my unscientific measurment, trial got slightly faster on twisted test suite from 78.1 to 77.4s for the full test suite.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @fijal|
|-|-|

Thanks for this.

  1. Can you add a docstring for `_setWarningRegistryToNone`, including docs for the type of `d` and what will happen to it?
  1. Can you add a docstring for `test__collectWarnings_bug` explaining the feature being tested?
  1. Also can you rename the test method?  Maybe something like `test_collectWarningsChangeWhileIterating`.
  1. And add an extra blank line before that test as well.
  1. in the import changes to `test_warning.py`, we prefer `(\n...)` over \-style continuations

Otherwise looks good, but I'd suggest checking the next version of the fix into a branch and forcing a build on buildbot before applying to trunk.

**Comment by twisted-trac** [comment]
|[<img alt="fijal's avatar" src="https://avatars.githubusercontent.com/u/85944?s=50" width="50" height="50">](https://github.com/fijal)<a name="note_8"></a>|@fijal removed owner|
|-|-|

Can't create a branch right now, lost my privileges somehow.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [32102]) Branching to 'collectwarnings-dict-iteration-4748'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @exarkun|
|-|-|

Thanks!

  1. Some coding standard things:
    1. docstring text should start on the line following the triple quote
    1. lines shouldn't have trailing whitespace
    1. 
  1. Other minor stuff
    1. `d` isn't the best variable name, I think `modules` would be a better name for the argument to `_setWarningRegistryToNone`.
    1. `ivar` in `_setWarningRegistryToNone` should be `param`
    1. Typo in first line of `_setWarningRegistryToNone` docstring - _ever_ should be _every_
  1. in the test
    1. Test method docstrings should just describe the behavior being tested without saying "Test that ...".
    1. Also stick to what you know :)  The py.test example is sufficient, I'm sure there are others, but we don't need to mention them (or we should mention them by name).

These are all minor, I'll fixed and apply.  Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [32104]) Minor coding-standard changes; make the unit test assert something.

refs [#4748](https://github.com/twisted/twisted/issues/4748)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [32105]) more clear news fragment

refs [#4748](https://github.com/twisted/twisted/issues/4748)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set status to `closed`|
|-|-|

(In [32106]) Merge collectwarnings-dict-iteration-4748

Author: fijal
Reviewer: exarkun
Fixes: [#4748](https://github.com/twisted/twisted/issues/4748)

When resetting `__warningregistry__` attributes on modules in trial,
iterate over a copy of the values of `sys.modules` so that if modules
are added or removed, trial can continue rather than being interrupted
by a "dictionary changed size during iteration" exception.
